File: test/domain/client_status_test.dart
import 'package:coup_laine/domain/models/animal_count.dart';
import 'package:coup_laine/domain/models/client.dart';
import 'package:coup_laine/domain/models/coordinates.dart';
import 'package:coup_laine/domain/use_cases/client_status.dart';
import 'package:flutter_test/flutter_test.dart';

Client _client({
  bool isWaiting = false,
  bool isBanned = false,
  List<AnimalCount> animals = const [AnimalCount(categoryId: 1, count: 5)],
}) {
  return Client(
    id: 1,
    name: 'X',
    addressLabel: '1 rue X',
    postcode: '22000',
    city: 'Saint-Brieuc',
    coordinates: const Coordinates(lat: 48, lon: -3),
    animals: animals,
    isWaiting: isWaiting,
    isBanned: isBanned,
  );
}

ClientStatus _derive(
  Client c, {
  bool planned = false,
  bool completed = false,
}) =>
    deriveStatus(
      c,
      hasPlannedTourThisSeason: planned,
      hasCompletedTourThisSeason: completed,
    );

void main() {
  group('deriveStatus priority', () {
    test('default when no flag and no tour', () {
      expect(_derive(_client()), ClientStatus.defaultStatus);
    });

    test('waiting flag → waiting', () {
      expect(_derive(_client(isWaiting: true)), ClientStatus.waiting);
    });

    test('planned tour beats waiting', () {
      expect(
        _derive(_client(isWaiting: true), planned: true),
        ClientStatus.scheduled,
      );
    });

    test('planned tour beats completed (scheduled surfaces upcoming work)',
        () {
      expect(
        _derive(_client(isWaiting: true), planned: true, completed: true),
        ClientStatus.scheduled,
      );
    });

    test('completed tour without planned → done', () {
      expect(
        _derive(_client(), completed: true),
        ClientStatus.done,
      );
    });

    test('animals empty → noAnimals', () {
      expect(
        _derive(
          _client(animals: const []),
          planned: true,
          completed: true,
        ),
        ClientStatus.noAnimals,
      );
    });

    test('animalsTotal > 0 → NOT noAnimals', () {
      // Any non-zero total keeps the client active.
      expect(
        _derive(_client(animals: const [AnimalCount(categoryId: 1, count: 3)])),
        ClientStatus.defaultStatus,
      );
    });

    test('banned beats everything', () {
      expect(
        _derive(
          _client(
            isBanned: true,
            animals: const [],
            isWaiting: true,
          ),
          planned: true,
          completed: true,
        ),
        ClientStatus.banned,
      );
    });
  });
}

File: tests/domain/use-cases/client-status.test.ts
import { describe, it, expect } from 'vitest';
import { computeClientStatus } from '@/domain/use-cases/client-status';

const SEASON = '2026-05-01';

describe('computeClientStatus (6-status)', () => {
  it('returns "banned" when isBanned is true (overrides everything)', () => {
    expect(
      computeClientStatus({
        isBanned: true,
        isWaiting: true,
        animalsTotal: 12,
        seasonStartedAt: SEASON,
        completedTourDates: ['2026-05-15'],
        plannedTourDates: ['2026-06-01'],
      })
    ).toBe('banned');
  });

  it('returns "noAnimals" when animalsTotal is 0 (and not banned)', () => {
    expect(
      computeClientStatus({
        isBanned: false,
        isWaiting: false,
        animalsTotal: 0,
        seasonStartedAt: SEASON,
        completedTourDates: [],
        plannedTourDates: [],
      })
    ).toBe('noAnimals');
  });

  it('returns "done" if any completed tour date ≥ seasonStartedAt', () => {
    expect(
      computeClientStatus({
        isBanned: false,
        isWaiting: true, // overridden by done
        animalsTotal: 12,
        seasonStartedAt: SEASON,
        completedTourDates: ['2026-05-15'],
        plannedTourDates: [],
      })
    ).toBe('done');
  });

  it('does not return "done" for completed tours before seasonStartedAt', () => {
    expect(
      computeClientStatus({
        isBanned: false,
        isWaiting: true,
        animalsTotal: 12,
        seasonStartedAt: SEASON,
        completedTourDates: ['2025-04-15'], // last season
        plannedTourDates: [],
      })
    ).toBe('waiting'); // not done; isWaiting wins
  });

  it('returns "scheduled" if a planned tour for this season exists and not done', () => {
    expect(
      computeClientStatus({
        isBanned: false,
        isWaiting: false,
        animalsTotal: 12,
        seasonStartedAt: SEASON,
        completedTourDates: [],
        plannedTourDates: ['2026-06-01'],
      })
    ).toBe('scheduled');
  });

  it('returns "waiting" when isWaiting and no scheduled/done', () => {
    expect(
      computeClientStatus({
        isBanned: false,
        isWaiting: true,
        animalsTotal: 12,
        seasonStartedAt: SEASON,
        completedTourDates: [],
        plannedTourDates: [],
      })
    ).toBe('waiting');
  });

  it('returns "default" when no other condition applies', () => {
    expect(
      computeClientStatus({
        isBanned: false,
        isWaiting: false,
        animalsTotal: 12,
        seasonStartedAt: SEASON,
        completedTourDates: [],
        plannedTourDates: [],
      })
    ).toBe('default');
  });
});
